### PR TITLE
One shot MFAC pruning support

### PR DESCRIPTION
**Pull request by bfineran**
One shot pruning support for WoodFisher / MFAC. This requires extra support because the method needs a certain number of gradient samples to prune with.  Support is added by offering `dataloader` and `loss_fn` kwargs that will be used to obtain the required samples.

Testing with dummy data/loss function passes through the MFAC pipeline:
Recipe contains - 
```yaml
  - !MFACGlobalPruningModifier
    params: "__ALL_PRUNABLE__"
    init_sparsity: 0.0
    final_sparsity: *target_sparsity
    start_epoch: *pruning_start_epoch
    end_epoch: *pruning_end_epoch
    update_frequency: *pruning_update_frequency
    mask_type: *pruning_mask_type
```

Sample code - 
```python
from sparseml.pytorch.optim import ScheduledModifierManager
from sparseml.pytorch.utils import GradSampler
import torch
from torchvision.models import resnet50

RECIPE_PATH = ""

r = resnet50(False)
manager = ScheduledModifierManager.from_yaml(RECIPE_PATH)

dataloader = ((torch.randn(10,3,224,224), 0) for _ in range(200))
loss_fn = lambda x, _: torch.mean(x)
grad_sampler = GradSampler(dataloader, loss_fn)

manager.apply(r, grad_sampler=grad_sampler)
```

### CODE REVIEW

File: src/sparseml/pytorch/utils/mfac_helpers.py [hunk 1]
```diff
@@ -38,6 +50,75 @@
...
+
+    def __init__(
+        self,
+        dataloader: Iterator[Tuple[Union[Tensor, List[Tensor]], Any]],
+        loss_fn: Callable[[Tensor], Tensor],
+    ):
+        self._dataloader = dataloader
+        self._loss_fn = loss_fn
+
+    @staticmethod
```
Comment (markurtz) [inline src/sparseml/pytorch/utils/mfac_helpers.py:73]: static method won't allow the implementer to override -- should probably change out for class or instance if we expect there's any chance they'll need to access state

File: src/sparseml/pytorch/utils/mfac_helpers.py [hunk 2]
```diff
@@ -38,6 +50,75 @@
...
+    def module_forward(module: Module, data: Union[Tensor, List[Tensor]]) -> Any:
+        """
+        :param module: module to perform forward pass with
+        :param data: single data sample to pass to module
+        :return: output(s) of the module forward pass
+        """
+        if isinstance(data, Tensor):
```
Comment (markurtz) [inline src/sparseml/pytorch/utils/mfac_helpers.py:80]: would it make sense to use the tensors_module_forward function here ideally making it more robust and generic?

File: src/sparseml/pytorch/utils/mfac_helpers.py [hunk 3]
```diff
@@ -38,6 +50,75 @@
...
+        each step
+    """
+
+    def __init__(
+        self,
+        dataloader: Iterator[Tuple[Union[Tensor, List[Tensor]], Any]],
+        loss_fn: Callable[[Tensor], Tensor],
+    ):
+        self._dataloader = dataloader
+        self._loss_fn = loss_fn
```
Comment (markurtz) [inline src/sparseml/pytorch/utils/mfac_helpers.py:71]: NIT: check if loss_fn is callable and dataloader is iterable

**Review by markurtz — CHANGES_REQUESTED** [review]

**Review by markurtz — APPROVED** [review]
LGTM

**Review by rahul-tuli — APPROVED** [review]
LGTM!
